File: workshop/processor/dotnet/Processor.cs
using System;
using System.IO;
using System.Collections.Concurrent;
using StackExchange.Redis;
using Microsoft.Extensions.CommandLineUtils;

namespace ConsoleApplication
{
    public class Processor
    {
        public string logfile { get; set; }
        public int threshold { get; set; }
        public ConcurrentDictionary<string, int> entries;
        public Processor(string logfile, int threshold)
        {
            this.logfile = logfile;
            this.threshold = threshold;
            this.entries = new ConcurrentDictionary<string, int>();
        }

        public string[] ReadFile()
        {
            return File.ReadAllLines(this.logfile);
        }

        public void Run()
        {
            foreach (var line in ReadFile())
            {   
                var parts  = line.Split(' ');
                var address = parts[0];
                var method = parts[5].Replace("\"", "");
                var uri = parts[6];                
                var responseCode = parts[8];

                if (method == "POST" && responseCode == "200")
                {
                    this.entries.AddOrUpdate(address, 1, (k,v) => v + 1);
                }
            }

            if (this.entries.Count > 0)
            {
                var connetion = ConnectionMultiplexer.Connect("127.0.0.1");
                IDatabase redis = connetion.GetDatabase();
                foreach (var entry in this.entries)
                {
                    Console.WriteLine("Blacklisting {0}. Threshold {1}, Actual {2}", entry.Key, this.threshold, entry.Value);
                    var key = entry.Key + ":repsheet:ip:blacklisted";
                    redis.StringSet(key, "Failed login processor");
                }
            }
        }
        public static void Main(string[] args)
        {
            CommandLineApplication cli = new CommandLineApplication(throwOnUnexpectedArg: false);
            CommandOption logFile = cli.Option("-f | --file <filename>", "The path and name of the logfile", CommandOptionType.SingleValue);
            CommandOption threshold = cli.Option(" -t | --threshold", "The blacklist threshold", CommandOptionType.SingleValue);
            cli.OnExecute(() => 
            {
                if (logFile.HasValue() && threshold.HasValue())
                {
                    var processor = new Processor(logFile.Value(), Int32.Parse(threshold.Value()));
                    processor.Run();
                }
                else
                {
                    Console.WriteLine("Must supply a logfile and threshold");
                }

                return 0;
            });
            cli.Execute(args);
        }
    }
}

File: workshop/processor/java/src/main/java/Processor.java
import redis.clients.jedis.Jedis;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.stream.Stream;

public class Processor {
    public static HashMap<String, Actor> counts  = new HashMap<>();
    private final String logFile;
    private static long threshold;
    private static Jedis jedis;

    public Processor(String logFile, long threshold) {
        this.logFile = logFile;
        Processor.threshold = threshold;
        jedis = new Jedis("localhost");
    }

    public static LogEvent getData(String line) {
        String[] parts = line.split(" ");
        return new LogEvent(parts[0], parts[5].replace("\"", ""), parts[8], parts[6]);
    }

    public static void updateCounts(LogEvent event) {
        Actor actor = counts.get(event.getAddress());

        if (actor == null) {
            actor = new Actor(event.getAddress());
        }

        if (event.getMethod().equals("POST") && event.getResponse().equals("200") && event.getUri().equals("/")) {
            actor.incrementFailedLogins();
            counts.put(event.getAddress(), actor);
        }

        if (Integer.parseInt(event.getResponse()) >= 400) {
            actor.incrementFailedResponse(event.getResponse());
            counts.put(event.getAddress(), actor);
        }
    }

    public static void check(String address, Actor actor) {
        if (actor.getFailedLogins() >= threshold) {
            System.out.println("Blacklisting " + actor.getName() + ". Failed Login Threshold 10, Actual: " + actor.getFailedLogins());
            jedis.set(actor.getName() + ":repsheet:ip:blacklisted", "failedLogin");
        }

        if (actor.getTotalFailedResponses() >= threshold) {
            System.out.println("Blacklisting " + actor.getName() + ". Failed Response Threshold 10, Actual: " + actor.getTotalFailedResponses());
            jedis.set(actor.getName() + ":repsheet:ip:blacklisted", "failedResponse");
        }
    }

    public void process() throws IOException {
        Path path = Paths.get(logFile);

        try (Stream<String> lines = Files.lines(path)) {
            lines.map(Processor::getData).forEach(Processor::updateCounts);
        }

        counts.forEach(Processor::check);
    }

    public void cleanup() {
        jedis.close();
    }

    public static void main(String[] args) throws IOException {
	Processor processor = new Processor(args[0], Long.parseLong(args[1]));
	processor.process();
	processor.cleanup();
    }
}
